### ISSUE TITLE
MacOS 3.00+ nightlies have an extended quit time

### ISSUE THREAD

**Issue by GScottElshaw** [transmission/transmission#2705]
It's often taking over 10 seconds after quitting Transmission nightlies for them to actually quit, often accompanied by the dreaded beach-ball as it finally terminates.

**Comment by ckerr** [comment]
Quick courtesy pong that I will look into this, but won't be able to until next week

**Comment by robd003** [comment]
@GScottElshaw how many torrents are active? How many different trackers are being used?

**Comment by GScottElshaw** [comment]
Currently 5 active, and around 10 -12 trackers.

**Comment by GScottElshaw** [comment]
@ckerr This last couple of nightly builds are significantly faster during quit. This current build is 3.00+ (66638cc80c).

**Comment by ckerr** [comment]
I'm seeing this too. From some quick checking it looks like Transmission is either waiting for libcurl to finish a task and/or not noticing that libcurl has finished, but that's as far as I've gotten so far.

**Comment by GScottElshaw** [comment]
I haven't been able to consistently replicate this at all. Sometimes fine, other times an age :-(

**Comment by GaryElshaw** [comment]
I updated cURL to 7.84.0 today and the builds i have been making with it seem to have improved the exit time, if others want to check.

**Comment by Coeur** [comment]
Is that something that can be addressed with a PR, @GaryElshaw ?
I'm definitely facing 10+ seconds quit durations. It's sometimes faster to build the whole code of the project than to quit Transmission.

**Comment by GaryElshaw** [comment]
> Is that something that can be addressed with a PR, <snip> ? 

I don't think so, unfortunately, but it's something i'll keep in mind to mention in the notes with regard to performance.

>I'm definitely facing 10+ seconds quit durations. It's sometimes faster to build the whole code of the project than to quit Transmission.

I was having that too, it would be 10 seconds or longer for it to quit. Not anymore. My version is courtesy of homebrew, but however you like to roll with it, i seriously recommend the update.

**Comment by Coeur** [comment]
I didn't see any perf difference between curl 7.79.1 and curl 7.84.0 regarding Transmission. Always takes 15+ sec to close.

**Comment by nevack** [comment]
Aren't we are linking with system libcurl on macOS?

If so, we need to find bottlenecks somewhere else, as libcurl path is pretty much is a dead end without bundling custom curl version with the app.

**Comment by Coeur** [comment]
Let's analyze the process itself:
1. Run with Xcode
2. Starting closing Transmission (cmd+Q)
3. Hit the break button in Xcode

This will reveal what is happening when closing.
![Capture d’écran 2022-07-05 à 19 24 06](https://user-images.githubusercontent.com/839992/177317288-c44ab60c-98e6-45a9-86d7-cb106e85720f.png)

So yes, my thread 8 has a loop with a 1 second `curl_multi_wait`, but my thread 6 seems more suspicious with a lot of UTP write buffer activity. When closing the app, we should interrupt UTP processing.

**Comment by GaryElshaw** [comment]
I don't think this solves the issue (as Antoine has pointed out), but exit time has drastically improved for me from what it was.

**Comment by GaryElshaw** [comment]
> we should interrupt UTP processing

I hope this might help with that #2996

**Comment by ckerr** [comment]
At least in my testing, the main issue is that Transmission is waiting for all the libcurl calls to finish up. It seems like there's always some tracker that doesn't respond quickly and that keeps Transmission waiting.

> I don't think this solves the issue (as Antoine has pointed out), but exit time has drastically improved for me from what it was.

That's very cool but I'm at a loss to explain why it would be better now. :confused: I don't _think_ it has anything to do with Transmission itself. **If** you were seeing the same "waiting on libcurl" issue that I'm seeing in my tests, *then* maybe things are better just because you're getting luckier than me with your choice of trackers :smile_cat:
